### PR TITLE
feat(new_metrics): migrate replica-level metrics for pegasus_server_impl (part 1)

### PR DESCRIPTION
**Pull request by empiredan**
This PR is to migrate replica-level metrics of `pegasus_server_impl` to new framework
(https://github.com/apache/incubator-pegasus/issues/1333). Since there are many replica-level
metrics in `pegasus_server_impl`, this PR is part 1 for this migration and other metrics (all
of which are gauges) will be migrated in the later PRs.

### CODE REVIEW

File: src/server/pegasus_server_impl.cpp [hunk 1]
```diff
@@ -670,41 +673,44 @@ void pegasus_server_impl::on_multi_get(multi_get_rpc rpc)
...
+            if (check_if_record_expired(epoch_now, value)) {
+                expire_count++;
+                if (_verbose_log) {
+                    LOG_ERROR_PREFIX(
+                        "rocksdb data expired for multi_get from {}: hash_key = \"{}\", "
+                        "sort_key = \"{}\"",
+                        rpc.remote_address(),
+                        ::pegasus::utils::c_escape_string(request.hash_key),
+                        ::pegasus::utils::c_escape_string(request.sort_keys[i]));
                 }
```
Comment (acelyc111) [inline src/server/pegasus_server_impl.cpp:696]: Seems duplicated several times, how about use another macro to wrap these code?

File: src/server/pegasus_server_impl.cpp [hunk 2]
```diff
@@ -670,41 +673,44 @@ void pegasus_server_impl::on_multi_get(multi_get_rpc rpc)
...
+                status = rocksdb::Status::NotFound();
+                continue;
             }
-            // if error occurred
-            if (!status.ok() && !status.IsNotFound()) {
-                error_occurred = true;
-                final_status = status;
+
+            // check if exceed limit
+            if (count >= max_kv_count || size >= max_kv_size) {
```
Comment (acelyc111) [inline src/server/pegasus_server_impl.cpp:702]: How about add `dsn_likely/unlikey` in hot paths?

File: src/server/pegasus_server_impl.cpp [hunk 3]
```diff
@@ -748,25 +754,24 @@ void pegasus_server_impl::on_multi_get(multi_get_rpc rpc)
...
     }
 
     if (expire_count > 0) {
-        _pfc_recent_expire_count->add(expire_count);
+        METRIC_VAR_INCREMENT_BY(read_expired_values, expire_count);
     }
     if (filter_count > 0) {
-        _pfc_recent_filter_count->add(filter_count);
+        METRIC_VAR_INCREMENT_BY(read_filtered_values, filter_count);
     }
```
Comment (acelyc111) [inline src/server/pegasus_server_impl.cpp:765]: How about move the judgement into METRIC_VAR_INCREMENT_BY or counter metric?

File: src/server/pegasus_write_service_impl.h [hunk 4]
```diff
@@ -689,7 +691,7 @@ class pegasus_write_service::impl : public dsn::replication::replica_base
     const std::string _primary_address;
     const uint32_t _pegasus_data_version;
 
-    ::dsn::perf_counter_wrapper &_pfc_recent_expire_count;
+    METRIC_VAR_DECLARE_counter(read_expired_values);
```
Comment (acelyc111) [inline src/server/pegasus_write_service_impl.h:694]: Anywhere we use it?

File: src/server/pegasus_server_impl.cpp (hunk 1 above)
Comment (empiredan) [reply src/server/pegasus_server_impl.cpp:696]: OK, I'll encapsulate this into functions naming `log_expired_data*`.

File: src/server/pegasus_server_impl.cpp (hunk 2 above)
Comment (empiredan) [reply src/server/pegasus_server_impl.cpp:702]: OK, add `dsn_unlikely` here.

File: src/server/pegasus_server_impl.cpp (hunk 3 above)
Comment (empiredan) [reply src/server/pegasus_server_impl.cpp:765]: Check if it is zero in `METRIC_VAR_INCREMENT_BY`.

File: src/server/pegasus_write_service_impl.h (hunk 4 above)
Comment (empiredan) [reply src/server/pegasus_write_service_impl.h:694]: This variable can be dropped.

File: src/server/pegasus_server_impl.cpp [hunk 5]
```diff
@@ -277,21 +334,20 @@ void pegasus_server_impl::on_get(get_rpc rpc)
...
     rocksdb::Status status = _db->Get(_data_cf_rd_opts, _data_cf, skey, &value);
 
     if (status.ok()) {
         if (check_if_record_expired(utils::epoch_now(), value)) {
-            _pfc_recent_expire_count->increment();
-            if (_verbose_log) {
-                LOG_ERROR_PREFIX("rocksdb data expired for get from {}", rpc.remote_address());
-            }
+            METRIC_VAR_INCREMENT(read_expired_values);
+            log_expired_data_if_verbose(__FUNCTION__, rpc.remote_address(), key);
```
[the remark this answers is not in the document]
Comment (empiredan) [reply src/server/pegasus_server_impl.cpp]: OK, I'll make it more graceful with macro.

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**